From f92ac0310cc20caf2c402ebd43b12c75ae33f96f Mon Sep 17 00:00:00 2001
From: Kirikaze Chiyuki <me@chyk.ink>
Date: Sun, 24 Nov 2024 10:01:08 +0800
Subject: [PATCH] Fix ln8000 charger and pm8150 real-time clock

These fixes were not belong to me, but I don't know the author's email address, so I'll commit as me on this repository temporarily.
Credits:
- ln8000 fixed by Timofey
- pm8150 rtc fixed by Catmengi
---
 arch/arm64/boot/dts/qcom/pm8150.dtsi          | 10 ++++-
 .../boot/dts/qcom/sm8150-xiaomi-nabu.dts      | 18 ++++++--
 drivers/power/supply/ln8000_charger.c         | 41 ++++++++++---------
 3 files changed, 46 insertions(+), 23 deletions(-)

diff --git a/arch/arm64/boot/dts/qcom/pm8150.dtsi b/arch/arm64/boot/dts/qcom/pm8150.dtsi
index a74a7ff660d2..51a745045eb0 100644
--- a/arch/arm64/boot/dts/qcom/pm8150.dtsi
+++ b/arch/arm64/boot/dts/qcom/pm8150.dtsi
@@ -118,13 +118,21 @@ pm8150_adc_tm: adc-tm@3500 {
 			status = "disabled";
 		};
 
-		rtc@6000 {
+		pm8150_rtc: rtc@6000 {
 			compatible = "qcom,pm8941-rtc";
 			reg = <0x6000>, <0x6100>;
 			reg-names = "rtc", "alarm";
 			interrupts = <0x0 0x61 0x1 IRQ_TYPE_NONE>;
 		};
 
+		pm8150_sdam_2: sdam@b100 {
+			compatible = "qcom,spmi-sdam";
+			reg = <0xb100>;
+			#address-cells = <1>;
+			#size-cells = <1>;
+			status = "disabled";
+		};
+
 		pm8150_gpios: gpio@c000 {
 			compatible = "qcom,pm8150-gpio", "qcom,spmi-gpio";
 			reg = <0xc000>;
diff --git a/arch/arm64/boot/dts/qcom/sm8150-xiaomi-nabu.dts b/arch/arm64/boot/dts/qcom/sm8150-xiaomi-nabu.dts
index db137bb583fd..e4d692b66bf0 100644
--- a/arch/arm64/boot/dts/qcom/sm8150-xiaomi-nabu.dts
+++ b/arch/arm64/boot/dts/qcom/sm8150-xiaomi-nabu.dts
@@ -506,7 +506,7 @@ ln8000_charger@51 {
 		compatible = "lionsemi,ln8000";
 		reg = <0x51>;
 
-		status = "disabled";
+		status = "okay";
 
 		irq-gpios = <&tlmm 36 0x2002>;
 
@@ -708,7 +708,7 @@ PDO_FIXED_USB_COMM |
 					 PDO_FIXED_DUAL_ROLE |
 					 PDO_FIXED_USB_COMM |
 					 PDO_FIXED_DATA_SWAP)
-					 PDO_VAR(5000, 12000, 3000)>;
+					 PDO_VAR(5000, 9000, 3000)>;
 
 		op-sink-microwatt = <10000000>;
 
@@ -811,7 +811,7 @@ &remoteproc_mpss {
 };
 
 &remoteproc_slpi {
- status = "okay";
+ status = "disabled";
  firmware-name = "qcom/sm8150/xiaomi/nabu/slpi_nb.mbn";
 };
 
@@ -1146,3 +1146,15 @@ &wifi {
 	vdd-1.3-rfa-supply = <&vreg_l2c_1p3>;
 	vdd-3.3-ch0-supply = <&vreg_l11c_3p3>;
 };
+
+&pm8150_rtc {
+	nvmem-cells = <&rtc_offset>;
+	nvmem-cell-names = "offset";
+	status = "okay";
+};
+&pm8150_sdam_2 {
+	status = "okay";
+	rtc_offset: rtc-offset@bc {
+		reg = <0xbc 0x4>;
+	};
+};
diff --git a/drivers/power/supply/ln8000_charger.c b/drivers/power/supply/ln8000_charger.c
index 1eabea38d718..8a2e26d8781b 100644
--- a/drivers/power/supply/ln8000_charger.c
+++ b/drivers/power/supply/ln8000_charger.c
@@ -1566,6 +1566,20 @@ static int ln8000_probe(struct i2c_client *client)
         dev_err(&client->dev, "%s: fail to alloc devm for ln8000_info\n", __func__);
         return -ENOMEM;
     }
+    info->dev = &client->dev;
+    info->client = client;
+
+    info->typec_psy = power_supply_get_by_phandle(info->dev->of_node,"usb-tcpm");
+
+    if(info->typec_psy == NULL){
+	ret = -(EPROBE_DEFER);
+	return ret;
+    }
+    if (IS_ERR(info->typec_psy)) {
+	    ret = PTR_ERR(info->typec_psy);
+	    dev_warn(info->dev, "Failed to get USB Type-C: %d\n", ret);
+	    info->typec_psy = NULL;
+    }
 
     info->pdata = devm_kzalloc(&client->dev, sizeof(struct ln8000_platform_data), GFP_KERNEL);
     if (info->pdata == NULL) {
@@ -1573,8 +1587,6 @@ static int ln8000_probe(struct i2c_client *client)
        kfree(info);
        return -ENOMEM;
     }
-    info->dev = &client->dev;
-    info->client = client;
     ret = ln8000_parse_dt(info);
     if (IS_ERR_VALUE((unsigned long)ret)) {
         ln_err("fail to parsed dt\n");
@@ -1625,18 +1637,9 @@ static int ln8000_probe(struct i2c_client *client)
 
     determine_initial_status(info);
 
-    info->typec_psy = power_supply_get_by_phandle(info->dev->of_node,
-							"usb-tcpm");
-	if (IS_ERR(info->typec_psy)) {
-		ret = PTR_ERR(info->typec_psy);
-		dev_warn(info->dev, "Failed to get USB Type-C: %d\n", ret);
-		info->typec_psy = NULL;
-	}
-
 	if (info->typec_psy) {
-		INIT_DELAYED_WORK(&info->status_changed_work,
-			ln8000_status_changed_worker);
-        INIT_DELAYED_WORK(&info->charge_work, psy_chg_get_ti_alarm_status);
+		INIT_DELAYED_WORK(&info->status_changed_work,ln8000_status_changed_worker);
+		INIT_DELAYED_WORK(&info->charge_work, psy_chg_get_ti_alarm_status);
 
 		info->nb.notifier_call = ln8000_notifier_call;
 		ret = power_supply_reg_notifier(&info->nb);
@@ -1646,11 +1649,11 @@ static int ln8000_probe(struct i2c_client *client)
 			return ret;
 		}
 
-        if (info->volt_qual) {
-            ln_info("start charging on init\n");
-            psy_chg_set_charging_enable(info, true);
-            schedule_delayed_work(&info->charge_work, msecs_to_jiffies(0));
-        }
+		if (info->volt_qual) {
+			ln_info("start charging on init\n");
+			psy_chg_set_charging_enable(info, true);
+			schedule_delayed_work(&info->charge_work, msecs_to_jiffies(0));
+		}
 	}
 
     return 0;
@@ -1759,4 +1762,4 @@ module_i2c_driver(ln8000_driver);
 MODULE_AUTHOR("sungdae choi<sungdae@lionsemi.com>");
 MODULE_DESCRIPTION("LIONSEMI LN8000 charger driver");
 MODULE_LICENSE("GPL v2");
-MODULE_VERSION("0.3.0");
\ No newline at end of file
+MODULE_VERSION("0.3.0");
-- 
2.51.0

